### ISSUE TITLE
training encounter with NAN

### ISSUE THREAD

**Issue by lhao0301** [abhimanyudubey/confusion#4]
Thanks for your wonderful code! I apply your code to a plant data set and encounter with `nan` easily (after some iterations, about 1k iteration). Settings are as follows:
a) category is about 3k 
b) lambda=140
c) softmaxwithloss
d)  add an extra softmax layer(normalize to 0~1), followed with pair confusion
e) batchsize=24, iter_size=6
f) lr_policy is changed to multistep
In the initial iteration, loss seems to be decreasing. I have tried decay the lr.
Could you please give some suggestions about it?

**Comment by abhimanyudubey** [comment]
Hey, 

Can you try lowering the value of lambda? Since we experimented mostly with problems of up to 500 classes, lambda was never higher than 50. Perhaps for 3k classes it doesn't need such a high value of lambda. This in turn would probably lead to very large gradients somewhere in the training (after cross-entropy is sufficiently small) that moves the outputs to `nan`. 

I'd say start with `lambda=10` and then increase/decrease to see how performance varies. What network are you using and which codebase (caffe or pytorch), and with which optimizer (adam or SGD)?

**Comment by lhao0301** [comment]
ok. I would try to decay lambda.
I finetune resnet-50 on pretrained model in caffe.

**Comment by lhao0301** [comment]
I increase/decay lambda. Both go into NAN finally after different iterations. I try to add gradient clipping but cause a terrible accuracy.

**Comment by abhimanyudubey** [comment]
Can you share your prototxt?

**Comment by lhao0301** [comment]
Train_val prototxt is as follows:
[train_val.txt](https://github.com/abhimanyudubey/confusion/files/2460503/train_val.txt)

**Comment by abhimanyudubey** [comment]
Thanks. Can you also share the prototxt you generate while using confusion loss? Also, does this go to NaN even without the loss?

**Comment by lhao0301** [comment]
I have updated the train_val prototxt with confusion loss(the last uploaded prototxt is a mistake ). Without confusion loss, the training process acts normally.

**Comment by abhimanyudubey** [comment]
Just took a look at your prototxt. Since I don't have this dataset, I can't test it myself, but I'd give 2 suggestions - i) try with a larger batch-size (32, 64 or 128 if possible), and ii) try with \lambda = 0.01, 0.1, 1 and so on.

**Comment by lhao0301** [comment]
For limited gpu memory, I set batch-size to 24 and set iter_size to 4(so batch size is 24 * 4 actually). That lambda is less than 1 is what I have not tried. Thanks!
